mongo_orchestration/replica_sets.py

```diff
@@ -49,43 +50,64 @@ def __init__(self, rs_params):
         self.repl_id = rs_params.get('id', None) or str(uuid4())
         self._version = rs_params.get('version')
 
         self.sslParams = rs_params.get('sslParams', {})
         self.kwargs = {}
+        self.restart_required = self.login or self.auth_key
+        self.x509_extra_user = False
 
-        if not not self.sslParams:
+        if self.sslParams:
             self.kwargs['ssl'] = True
+            self.kwargs['ssl_certfile'] = DEFAULT_CLIENT_CERT
 
+        members = rs_params.get('members', {})
         config = {"_id": self.repl_id, "members": [
-                  self.member_create(member, index) for index, member in enumerate(rs_params.get('members', {}))
-                  ]}
+            self.member_create(member, index)
+            for index, member in enumerate(members)
+        ]}
         logger.debug("replica config: {config}".format(**locals()))
         if not self.repl_init(config):
             self.cleanup()
             raise ReplicaSetError("Could not create replica set.")
 
+        if not self.waiting_config_state():
+            raise ReplicaSetError(
+                "Could not actualize replica set configuration.")
+
         if self.login:
-            logger.debug("add admin user {login}/{password}".format(login=self.login, password=self.password))
-            try:
-                c = self.connection()
-                c.admin.add_user(self.login, self.password,
-                                 roles=['__system',
-                                        'clusterAdmin',
-                                        'dbAdminAnyDatabase',
-                                        'readWriteAnyDatabase',
-                                        'userAdminAnyDatabase'])
-                # Make sure user propagates to secondaries before proceeding.
-                c.admin.authenticate(self.login, self.password)
-                c.admin.command('getLastError', w=len(self.servers()))
-            except pymongo.errors.OperationFailure:
-                reraise(ReplicaSetError,
-                        "Could not add user %s to the replica set."
-                        % self.login)
-            finally:
-                c.close()
+            # Do we need to add an extra x509 user?
+            for member in members:
+                proc_params = member.get('procParams', {})
+                set_params = proc_params.get('setParameter', {})
+                auth_mechs = set_params.get('authenticationMechanisms', '')
+                auth_mechs = auth_mechs.split(',')
+                if len(auth_mechs) == 1 and auth_mechs[0] == 'MONGODB-X509':
+                    self.x509_extra_user = True
+                    break
+
+            self._add_users(self.connection()[self.auth_source])
+        if self.restart_required:
+            # Restart all the servers with auth flags and ssl.
+            for idx, member in enumerate(members):
+                server_id = self._servers.id_by_hostname(self.id2host(idx))
```

Comment:
Yes they are. How about cleaning that up? Maybe member_id_to_host, server_id_by_hostname?